qa: fix ltc-scrypt

xanimo opened the pull request:
-update install-deps.sh so it will clean up after itself and can be invoked from root directory while unpacking to correct location (qa/pull-tester/ltc-scrypt-master/).
-forked langerhans fork and updated ltc-scrypt-master to define PY_SSIZE_T_CLEAN macro prior to #include <Python.h> and updated scrypt_getpowhash variables inputlen and outbuflen to Py_ssize_t types.
-updated qa/README.md to include sudo prior to apt-get install directive and add instructions to invoke script from root dir.
-updated mininode.py with custom ripemd160 (otherwise bumpfee.py fails).

xanimo commented:
I'm not sure if @langerhans still maintains [ltc-scrypt-master](https://github.com/langerhans/ltc-scrypt). If so I'll gladly submit a PR to fix that and remove my fork from install-deps.sh. Either way this was meant to demonstrate a working solution now considering @patricklodder's comment on expeditiously resolving (urgency).

patricklodder commented:
> I'm not sure if @langerhans still maintains [ltc-scrypt-master](https://github.com/langerhans/ltc-scrypt). If so I'll gladly submit a PR to fix that and remove my fork from install-deps.sh

Now: https://github.com/dogecoin/ltc-scrypt

Big thx to @langerhans - please PR the changes there.

patricklodder commented:
I've marked this as draft until this integrates with the transferred repo.

patricklodder commented:
@xanimo: thinking of it, could you please PR the python-native ripemd160 separately? That's another OpenSSL3 related regression. That doesn't have to be blocked by the ltc-scrypt move we're doing.

xanimo commented:
Changing from draft to ready for review since it works albeit lacking a tagged release version of dogecoin/ltc-scrypt (can wait or update later 🤷). Revised PR description is:
```
-updated .gitignore with extracted archives path and install-deps.sh so it will clean up after itself and can be invoked from root directory while unpacking to correct location (qa/pull-tester/ltc-scrypt-master/).
-updated qa/README.md to include sudo prior to apt-get install directive and add instructions to invoke script from root directory.
```

patricklodder commented:
I have added a CI at dogecoin/ltc-scrypt#9 so that we have that out of the way and can tag it, so that this can be truly fixated - on a tag rather than a branch that may mutate. I'd suggest waiting for the tag.

patricklodder commented on qa/pull-tester/install-deps.sh:
```diff
@@ -1,10 +1,11 @@
+echo "e866ade37fb27439ae0ca32f1ee4ad32be428c1fdac9bcc988b36c68648ff0de  $file" | sha256sum -c
+tar zxf $file -C $workdir/ --transform 's/ltc-scrypt-1.0.1/ltc-scrypt/' && rm -rf $file
+pushd $workdir/ltc-scrypt
+python3 -m pip install .
```
Minor undocumented change here: the difference with this installation method (`pip install .`) and the previous one (`setup.py install --user`) is that this does not explicitly install into userspace but only defaults to it if the user that runs the script does not have write access to the system install path. Is it possible to make the userspace install explicit just in case?

xanimo replied:
Ah! Good point. I can affix with `--user` to explicitly enforce (`python3 -m pip install . --user`) but also don't want to presume folks have pip installed despite the setuptools deprecation warnings to shift to standards-based tools:
```
/mnt/source/repos/dogecoin/qa/pull-tester/ltc-scrypt /mnt/source/repos/dogecoin
running install
/home/bluezr/.local/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/home/bluezr/.local/lib/python3.10/site-packages/setuptools/command/easy_install.py:144: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/home/bluezr/.local/lib/python3.10/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 1.16.0-unknown is an invalid version and will not be supported in a future release
  warnings.warn(
/home/bluezr/.local/lib/python3.10/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 0.1.43ubuntu1 is an invalid version and will not be supported in a future release
  warnings.warn(
/home/bluezr/.local/lib/python3.10/site-packages/pkg_resources/__init__.py:123: PkgResourcesDeprecationWarning: 1.1build1 is an invalid version and will not be supported in a future release
```

Would that work or should I revert back? Seeing tests failing anyway so perhaps I should revert.

patricklodder replied:
I'd recommend it with the `--user` flag, as that would make us compatible with python 3.14+ too ❤️

The CI seems to fail on not being able to find distutils, can be fixed by adding `python3-distutils` from apt

xanimo replied:
I've appended the installation directive in install-deps.sh with `--user` flag and added `python3-pip` and `python3-setuptools` to the 3 that were previously failing in ci.yml and they now pass. 

I'll have to reconfirm but I also needed `python3-dev` instead of `python3` because the Python.h header was unavailable during installation when building manually in an i386/ubuntu:bionic container early last night. May just be able to install specifically for that as oppose to change for every matrix.host if that's preferred. Otherwise can wait till next iteration of ltc-scrypt to see if that will allow avoiding any amendments in ci.yml.

patricklodder commented on .github/workflows/ci.yml:
```diff
@@ -220,7 +220,7 @@ jobs:
         run: |
           sudo apt-get update
-          sudo apt-get install build-essential libtool autotools-dev automake pkg-config bsdmainutils curl ca-certificates ccache python3 rsync git procps bison
+          sudo apt-get install build-essential libtool autotools-dev automake pkg-config bsdmainutils curl ca-certificates ccache python3-dev rsync git procps bison
```
please only add `python3-dev` where we need it (in the matrix for those jobs where we run the `qa/` tests, because we shouldn't need it for anything else)

xanimo replied:
have reverted back to python3.

patricklodder commented on qa/README.md:
```diff
@@ -13,16 +13,17 @@ Unix
-sudo apt-get update && apt-get install -y curl python3-zmq python3-dev gcc
-cd qa/pull-tester/ && ./install-deps.sh
+sudo apt-get update
+sudo apt-get install -y curl python3-zmq python3-dev gcc
```
does this need `python3-setuptools` too?

xanimo replied:
i've updated the pertinent sections in ci that run qa to keep `python3-zmq` and install `python3-pip` as the recommended packages installed on `bionic` are:
```
root@2ad4e01d2056:~/dogecoin# apt-get install python3-pip
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  dh-python python3-asn1crypto python3-cffi-backend python3-crypto python3-cryptography python3-dbus
  python3-dev python3-distutils python3-gi python3-idna python3-keyring python3-keyrings.alt python3-lib2to3
  python3-pkg-resources python3-secretstorage python3-setuptools python3-six python3-wheel python3-xdg
Suggested packages:
  python-crypto-doc python-cryptography-doc python3-cryptography-vectors python-dbus-doc python3-dbus-dbg
  gnome-keyring libkf5wallet-bin gir1.2-gnomekeyring-1.0 python-secretstorage-doc python-setuptools-doc
The following NEW packages will be installed:
  dh-python python3-asn1crypto python3-cffi-backend python3-crypto python3-cryptography python3-dbus
  python3-dev python3-distutils python3-gi python3-idna python3-keyring python3-keyrings.alt python3-lib2to3
  python3-pip python3-pkg-resources python3-secretstorage python3-setuptools python3-six python3-wheel
  python3-xdg
```
will have to update this at `bionic` eol anyway but works locally for time being. waiting for ci to confirm, otherwise will add required packages explicitly.

xanimo replied:
hmm seems recommended packages were not installed so i'll have to add explicitly and re-run. feel free to cancel the ci run if you want as it's failed.

patricklodder commented on .gitignore:
```diff
@@ -103,6 +103,7 @@ linux-coverage-build
 linux-build
 win32-build
 qa/pull-tester/tests_config.py
+qa/pull-tester/ltc-scrypt
```
is this still needed now that we install from archive?

xanimo replied:
it is not. i have removed.

patricklodder reviewed (approved):
ACK - tested on macOS 12.5.1 and Ubuntu 22.04
